code/modules/paperwork/filingcabinet.dm

```diff
@@ -65,20 +65,46 @@
 	else
 		return ..()
 
+/obj/structure/filingcabinet/attack_hand(mob/living/carbon/user, list/modifiers)
+	. = ..()
+	ui_interact(user)
+
+/obj/structure/filingcabinet/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "FilingCabinet")
+		ui.open()
+
+/obj/structure/filingcabinet/ui_data(mob/user)
+	var/list/data = list()
+
+	data["cabinet_name"] = "[name]"
+	data["contents"] = list()
+	data["contents_ref"] = list()
+	for(var/obj/item/content in src)
+		data["contents"] += "[content]"
+		data["contents_ref"] += "[REF(content)]"
 
-/obj/structure/filingcabinet/ui_interact(mob/user)
+	return data
+
+
+/obj/structure/filingcabinet/ui_act(action, params)
 	. = ..()
-	if(contents.len <= 0)
-		to_chat(user, span_notice("[src] is empty."))
+	if(.)
+		return
+
+	if(usr.stat != CONSCIOUS || HAS_TRAIT(usr, TRAIT_HANDS_BLOCKED))
 		return
 
-	var/dat = "<center><table>"
-	var/i
-	for(i=contents.len, i>=1, i--)
-		var/obj/item/P = contents[i]
-		dat += "<tr><td><a href='?src=[REF(src)];retrieve=[REF(P)]'>[P.name]</a></td></tr>"
-	dat += "</table></center>"
-	user << browse("<html><head><meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><title>[name]</title></head><body>[dat]</body></html>", "window=filingcabinet;size=350x300")
+	switch(action)
+		// Take the object out
+		if("remove_object")
+			var/obj/item/content = locate(params["ref"]) in src
+			if(istype(content) && in_range(src, usr))
+				usr.put_in_hands(content)
+				updateUsrDialog()
```

Comment:
you really don't need this, just return TRUE at the end.

Reply:
I kept `updateUsrDialog()` from the old version of the code. Is it for the old ui?

Reply:
Do you mean `updateUsrDialog()` with *this*? Should I replace that with `return TRUE`? I don't know what that is or why it is there.